### PR TITLE
Support mixed v2/v3 xDS request/responses 

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
Commit Message: To better support migration from v2 type url to v3 type url, we should be able to support a DiscoveryResponse with a v3 resource to a DiscoveryRequest with v2 resource type. This can be achieved by downgrading the type url when subscriber can't be found.
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue][ Support mixed v2/v3 xDS request/responses](https://github.com/envoyproxy/envoy/issues/10776)
[Optional Deprecated:]

### CODE REVIEW

File: source/common/config/api_type_oracle.cc
Comment (Shikugawa) [inline source/common/config/api_type_oracle.cc:35]: Can't you use this? This may have duplication. https://github.com/envoyproxy/envoy/blob/1323ff23a9372cf5492ea567ea02cca54c891496/source/common/config/resource_name.h#L35

**Review by Shikugawa — COMMENTED** [review]
Thank you! Generally LGTM. I'd commented on some of the sections.

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc:35]: Thanks! I didn't know there is a url conversion interface out there.

Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc:35]: I can't use that template because I don't have the typename of the resources.

**Comment by htuch** [comment]
@adisuissa could you take a first pass? Thanks.

File: source/common/config/api_type_oracle.cc
Comment (dmitri-d) [inline source/common/config/api_type_oracle.cc:38]: Why not use string::find here?

File: source/common/config/api_type_oracle.cc
Comment (dmitri-d) [inline source/common/config/api_type_oracle.cc:54]: Wouldn't hurt to have a test for this, esp. if we keep the code to lookup for the "/" in the type url.

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc:38]: Yes, that makes sense. I will use string::find.

File: source/common/config/grpc_mux_impl.cc
Comment (adisuissa) [inline source/common/config/grpc_mux_impl.cc:128]: nit: pass `type_url` as a parameter to keep consistency in this scope

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:157]: Sorry, forgot to clean it up.

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:128]: Thanks, will do.

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc:43]: Yes, that make sense.

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc]: Thanks, that makes sense.

File: source/common/config/api_type_oracle.cc
Comment (dmitri-d) [inline source/common/config/api_type_oracle.cc:40]: I forgot that there's also absl::StrSplit() which might be more readable than this, I'll leave it up to you.

File: source/common/config/grpc_mux_impl.cc
Comment (dmitri-d) [inline source/common/config/grpc_mux_impl.cc:154]: nit: Here and the return above -- the call always returns from this conditional, perhaps move returns to right after it? Would make it slightly more readable.

**Review by dmitri-d — COMMENTED** [review]
lgtm other than nits

File: source/common/config/api_type_oracle.cc
Comment (htuch) [reply source/common/config/api_type_oracle.cc:35]: `TypeUtil::typeUrlToDescriptorFullName` is what you're after I think.

File: source/common/config/api_type_oracle.cc
Comment (htuch) [inline source/common/config/api_type_oracle.cc:44]: You probably want to generally work with type names in this file, and only at the point at which you are doing type URLs in the watcher etc. do the conversion to type URL. FWIW, I think we should have a `TypeUtil::descriptionFullNameToTypeurl` to factor out this low level manipulation.

File: source/common/config/grpc_mux_impl.cc
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc:126]: When do we hit the second conjunct? I would think in reality, we are generally subscribing to v3, so if v2 comes along, we hit the first conjunct.

File: source/common/config/grpc_mux_impl.cc
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc:127]: Nit: `const`

File: source/common/config/grpc_mux_impl.cc
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc:147]: Why this change?

File: test/integration/ads_integration_test.cc
Comment (htuch) [inline test/integration/ads_integration_test.cc:136]: This looks like a v3 type URL :)

File: source/common/config/grpc_mux_impl.cc
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc:132]: So, thinking back to our conversation the other day, we definitely don't need to worry about v4alpha, so v2/v3 is fine as you have it. But, one thing that might not be treated correctly here is that we could have an EDS subscription to v3, but a v2 type URL appears in the message. Does this need special treatment?

**Review by htuch — CHANGES_REQUESTED** [review]
Looks great, thanks, flushing a few comments.
/wait

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:126]: In ads integration test, for some resources, both v2 and v3 type url are watched, but v3 type url doesn't have any watcher.

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:132]: The issue doesn't cover the case where we have v2 response to v3 request, this can be fixed easily though.

Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:132]: But I guess we currently don't have any interface to convert a v2 type url to a v3 type url?

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:147]: That is unintentional, I will clean it up.

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:154]: Sure, will clean it up.

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:132]: We can always downgrade type url to v2 in addWatch(), which can solve this problem, but it will cause performance penalty for common use cases. For example, v2 response to v2 request, v3 response to v3 request.

File: source/common/config/new_grpc_mux_impl.cc
Comment (adisuissa) [inline source/common/config/new_grpc_mux_impl.cc:55]: nit: const

**Review by adisuissa — COMMENTED** [review]
Looks good, thanks!
A couple of nits.

**Comment by chaoqin-li1123** [comment]
I have added support for case where grpc_mux watch for v3 type_url but receive v2 type_url resource.

File: source/common/config/grpc_mux_impl.cc
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc:142]: Why remove this comment?

File: include/envoy/config/grpc_mux.h
Comment (chaoqin-li1123) [reply include/envoy/config/grpc_mux.h:106]: Sure, I will move it to impl file.

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc:36]: Thanks, will do.

File: source/common/config/new_grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/new_grpc_mux_impl.cc]: Thanks, will do.

File: include/envoy/config/grpc_mux.h
Comment (Shikugawa) [inline include/envoy/config/grpc_mux.h:105]: Should we put this on interface? I think that this information should not have here.

Comment (chaoqin-li1123) [reply include/envoy/config/grpc_mux.h:105]: Sure, will move it to impl file.

Comment (htuch) [reply include/envoy/config/grpc_mux.h:105]: Might make sense to move to an implementation file local `CONSTRUCT_ON_FIRST_USE` singleton, since it's only accesses on the main thread and the type mapping doesn't change.

Comment (chaoqin-li1123) [reply include/envoy/config/grpc_mux.h:105]: Yes, that makes sense.

File: source/common/config/grpc_mux_impl.h
Comment (Shikugawa) [inline source/common/config/grpc_mux_impl.h:41]: I'm not sure if `enable_type_url_downgrade_and_upgrade` is required. Is it not good just to use runtime flag? I think that we should allow fallback behavior whenever the runtime feature is enabled.

File: source/common/config/new_grpc_mux_impl.cc
Comment (Shikugawa) [inline source/common/config/new_grpc_mux_impl.cc:85]: nit: not needed, or we can use more clear message.

Comment (chaoqin-li1123) [reply source/common/config/new_grpc_mux_impl.cc]: Sorry, forgot to clean up. It is for debug...

File: source/common/config/grpc_mux_impl.h
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.h:41]: This constructor is for testing purpose, in unit test we don't have runtime flag.

**Comment by ramaraochavali** [comment]
@htuch Will this runtime support work beyond Dec 2020 ? Or is this valid till Dec 2020? The reason I am asking is let us say a controlplane currently supports v2 only, I understand that with this PR, we will be able to configure Envoy to send v3 requests which will then be downgraded and upgraded back to v3 - so management servers can continue to use v2 for some more time with whatever perf penalty?

Am I understanding this correctly?

**Comment by chaoqin-li1123** [comment]
This PR doesn't downgrade the discovery request and discovery response itself. It allows Envoy to receive discovery responses with resource version different from the discovery request. To make it works, the control plane should be able to recognize the type url used by Envoy, then the control plane can send back any version of resource as long as the type match.

**Comment by ramaraochavali** [comment]
Ok. So Envoy sends v3 and control plane understands that it is v3 and responds it with v2  - then this PR will upgrade the v2 response to v3 response?

My question about timing still remains.How long would Envoy support this run time flag?

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc]: I remove that condition now. I thought that watcher of the same type of resources can share resources. But they shouldn't.

**Comment by htuch** [comment]
@ramaraochavali v2 is currently scheduled for removal completely from Envoy after the start of 2021 Q1 release.

File: test/common/config/grpc_mux_impl_test.cc
Comment (htuch) [inline test/common/config/grpc_mux_impl_test.cc:771]: What role does this second block play in testing (same for test below)? All things being equal, shorter tests for same line/functional coverage are better.

**Review by htuch — CHANGES_REQUESTED** [review]
Looks great. If we can wrap up the unresolved threads, this is good to go.
/wait

**Comment by ramaraochavali** [comment]
@htuch  Sorry. My question is about support of this runtime flag that accepts v2 responses from control plane. Is there a timeline attached to its support?

File: test/integration/ads_integration_test.cc
Comment (chaoqin-li1123) [reply test/integration/ads_integration_test.cc]: Have fixed that.

File: source/common/config/grpc_mux_impl.h
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.h]: I have removed the constructor with bool set explicitly.

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:28]: Thanks, will do.

File: source/common/config/grpc_mux_impl.h
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.h:143]: Will clean up.

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:127]: Thanks.

File: source/common/config/grpc_mux_impl.cc
Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:146]: Thanks, will clean up.

Comment (chaoqin-li1123) [reply source/common/config/grpc_mux_impl.cc:146]: Can not mark it const here, because [] operator is not a const function.

File: source/common/runtime/runtime_features.cc
Comment (chaoqin-li1123) [reply source/common/runtime/runtime_features.cc:97]: thanks, will do.

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc]: I have moved this helper function to api_type_oracle.cc and factor out the low level manipulation.

File: test/common/config/grpc_mux_impl_test.cc
Comment (chaoqin-li1123) [reply test/common/config/grpc_mux_impl_test.cc]: Thanks, I have removed the second block.

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc]: I have figured out a way to use helper function.

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc:55]: Now I use TypeUtil::typeUrlToDescriptorFullName and add testing.

File: docs/root/faq/api/control_plane_version_support.rst
Comment (htuch) [inline docs/root/faq/api/control_plane_version_support.rst:33]: I think this belongs in a slightly different part of this advice. This enumerated list gives how to do a rollout with a management server that supports mixed v2/v3 versions. Your PR is adding support so that a control plane operator can rollout a v3-only management server in the presence of v2-only management servers and have the clients work with both. It's a more flexible way to do rollouts.

File: docs/root/faq/api/control_plane_version_support.rst
Comment (htuch) [inline docs/root/faq/api/control_plane_version_support.rst:36]: Still need to add an entry in the version history, see https://www.envoyproxy.io/docs/envoy/latest/version_history/current

File: source/common/config/api_type_oracle.cc
Comment (htuch) [inline source/common/config/api_type_oracle.cc:43]: The utilities are more general than just the API type oracle, hence I think they should be in their own standalone library in `source/common/protobuf`.

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo comments on version history, docs and cleaning up the type URL util location.
/wait

File: source/common/config/api_type_oracle.cc
Comment (chaoqin-li1123) [reply source/common/config/api_type_oracle.cc:43]: Yes, that sounds reasonable.

File: docs/root/faq/api/control_plane_version_support.rst
Comment (chaoqin-li1123) [reply docs/root/faq/api/control_plane_version_support.rst:36]: Thanks, will do.

File: docs/root/version_history/current.rst
Comment (htuch) [inline docs/root/version_history/current.rst:130]: Tiny nit: Use backticks 
```
`
```

instead of

```
'
```

when quoting.

File: source/common/config/utility.cc
Comment (htuch) [inline source/common/config/utility.cc:274]: I think the conversion utils belong in `source/common/protobuf`, not `source/common/config`, since they are about type URLs, and hence generic to all protobuf.

File: source/common/protobuf/type_util.cc
Comment (htuch) [inline source/common/protobuf/type_util.cc:4]: This shouldn't be in the config namespace, but instead in some protobuf prefixed one., e.g. `ProtobufTypeUril`

Comment (chaoqin-li1123) [reply source/common/protobuf/type_util.cc:4]: Other helper function and class in protobuf repo don't have a nested namespace because Protobuf namespace is reserved for standard Protobuf library. I will just remove the Config namespace.

File: source/common/config/utility.cc
Comment (chaoqin-li1123) [reply source/common/config/utility.cc:274]: Thanks, will remove the Config namespace.

File: docs/root/faq/api/control_plane_version_support.rst
Comment (htuch) [reply docs/root/faq/api/control_plane_version_support.rst:33]: @chaoqin-li1123 I think the entire two paragraphs you added belong in a separate section below, not in the list of steps for servers which both support v2 and v3. There is essentially guidance for servers with both v2/v3, which is steps 1-4, and another section for servers that only have v2 XOR v3 during a rollout.

File: test/common/protobuf/utility_test.cc
Comment (htuch) [inline test/common/protobuf/utility_test.cc:50]: Seems unrelated, but I guess as useful thing to add.

**Review by htuch — CHANGES_REQUESTED** [review]
Ready to ship once the RST docs suggestions is done, thanks!
/wait

File: docs/root/faq/api/control_plane_version_support.rst
Comment (chaoqin-li1123) [reply docs/root/faq/api/control_plane_version_support.rst:33]: Thanks, will change the doc.

File: test/common/protobuf/utility_test.cc
Comment (chaoqin-li1123) [reply test/common/protobuf/utility_test.cc:50]: I changed this to add test coverage. The change is no longer needed after the refactoring, will undo the change.

**Review by htuch — APPROVED** [review]
LGTM, this is great!

**Comment by htuch** [comment]
@ramaraochavali will be supported until immediately after Q1 2021 release, planning on removing v2 at that point.

**Pull request merged**
